oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/OneandoneComputeServiceAdapter.java

```diff
@@ -62,77 +61,95 @@
-
-      //prepare hdds to provision
-      List<? extends Volume> volumes = hardware.getVolumes();
-      List<Hdd.CreateHdd> hdds = new ArrayList<Hdd.CreateHdd>();
-
-      for (final Volume volume : volumes) {
-         try {
-            //check if the bootable device has enough size to run the appliance(image).
-            float minHddSize = volume.getSize();
-            if (volume.isBootDevice()) {
-               SingleServerAppliance appliance = api.serverApplianceApi().get(image.getId());
-               if (appliance.minHddSize() > volume.getSize()) {
-                  minHddSize = appliance.minHddSize();
-               }
-            }
-            Hdd.CreateHdd hdd = Hdd.CreateHdd.create(minHddSize, volume.isBootDevice());
-            hdds.add(hdd);
-         } catch (Exception ex) {
-            throw Throwables.propagate(ex);
-
+      String imageId = image.getId();
+      Hardware hardwareModel = generateHardwareRequest.isFlavor(hardware.getId());
+      boolean isBaremetal = hardware.getName() == null ? false : hardware.getName().contains(baremetalModelsKey);
+      ServerAppliance workingImage = null;
+
+      //choose the correct image based on the server type baremetal or cloud
+      if (hardwareModel != null) {
+         if (isBaremetal) {
+            workingImage = findWorkingImage(image.getId(), Types.ServerTypeCompatibility.baremetal);
+         } else {
+            workingImage = findWorkingImage(image.getId(), Types.ServerTypeCompatibility.cloud);
+         }
+         if (workingImage != null) {
+            imageId = workingImage.id();
+         }
+      } else {
+         //check if image is of Type IMAGE
+         workingImage = findWorkingImage(image.getId(), Types.ServerTypeCompatibility.cloud);
+         if (workingImage != null) {
+            imageId = workingImage.id();
          }
       }
 
-      // provision server
-      Server server = null;
-      Double cores = ComputeServiceUtils.getCores(hardware);
-      Double ram = (double) hardware.getRam();
-      if (ram < 1024) {
-         ram = 0.5;
-      } else {
-         ram = ram / 1024;
+      //configuring Firewall rules
+      Map<Integer, Integer> portsRange = getPortRangesFromList(inboundPorts);
+      List<FirewallPolicy.Rule.CreatePayload> rules = new ArrayList<FirewallPolicy.Rule.CreatePayload>();
+
+      for (Map.Entry<Integer, Integer> range : portsRange.entrySet()) {
+         FirewallPolicy.Rule.CreatePayload rule = FirewallPolicy.Rule.CreatePayload.builder()
+                 .portFrom(range.getKey())
+                 .portTo(range.getValue())
+                 .protocol(Types.RuleProtocol.TCP)
+                 .build();
+         rules.add(rule);
+      }
+      String firewallPolicyId = "";
+      if (inboundPorts.length > 0) {
+         FirewallPolicy firewallPolicy = api.firewallPolicyApi().create(FirewallPolicy.CreateFirewallPolicy.create(name + " firewall policy", "desc", rules));
```

Comment:
what happens when you ask to create multiple nodes? does `firewallPolicyApi` handles the concurrency correctly?

jclouds usually takes care of fw rules for a group of node extending `CreateNodesWithGroupEncodedIntoNameThenAddToSet` -- see `ApplyNovaTemplateOptionsCreateNodesWithGroupEncodedIntoNameThenAddToSet` in openstack-nova api for an example

Reply:
It does handle the creation correctly as all the live tests passes, if you think there is a necessary change here, we can do that in another PR that addresses improvements.

This bulk of code was moved up just to organize the process more. it has been merged before since the first PR

Reply:
I've now realised I already commented this section: I'm afraid you need to change the code now otherwise we can't merge this PR. The reason for that is to have homogenous code, easier to maintain and debug.
if each contributions start deviating from jclouds's best practices it would become impossible for maintainers to help out. I'm sure you understand